### ISSUE TITLE
RFC: Restrict Eleventy’s access to directories outside of its realm

### ISSUE THREAD

**Issue by kleinfreund** [11ty/eleventy#368]
Currently, Eleventy does not restrict access to directories outside of its realm. This means that both input and output directory can be outside of Eleventy’s project directory (i.e. the place from where you run `eleventy`, a.k.a. the command directory). The following is possible:

```sh
eleventy --input=../private --output=../other-project
```

This is dangerous and a potential security risk for the following reasons:

- Users might override their own data by accident.
- In a hosted environment (see [jekyllrb.com: Security Fixes for series 3.6, 3.7 and 3.8](https://jekyllrb.com/news/2018/09/19/security-fixes-for-3-6-3-7-3-8/)), this would be an immediate non-starter.

I propose disallowing Eleventy from reading/writing outside of its root/project/command directory.

**Comment by zachleat** [comment]
yeah, I’m on board with this. `--input` is an easier sell than `--output` I think. But I’m on board with both.

Would this also require a review of any templating `include` functionality?

**Comment by kleinfreund** [comment]
Algorithm:

1. Normalize and strip leading double-dot segments from path
2. Use [Node.js’ fs.realpath](https://nodejs.org/api/fs.html#fs_fs_realpath_path_options_callback) to check …

   - if the path doesn’t exist, throw an error.
   - if path is outside of Eleventy’s realm, throw an error.

   Otherwise, return the absolute path.

This algorithm allows usage of symbolic links as long as they point inside Eleventy’s realm while restricting access to any directory outside of it.

Node.js implementation (draft) for step 2:

```node
const fs = require('fs');

function safePath(path) {
  try {
    const realPath = fs.realpathSync(path);

    if (realPath.startsWith(process.cwd())) {
      return path;
    }

    throw new Error(`Path ${realPath} is outside of Eleventy’s project directory.`);
  } catch (error) {
    throw new Error(error.path ? `Path ${error.path} does not exist.` : error);
  }
}
```

**Comment by zachleat** [comment]
@kleinfreund `_includes` should be completely optional unless you try to use it for something.

**Comment by aral** [comment]
Unless I’m mistaken, this change will make it much harder (if not impossible) to include eleventy in any other command-line apps (e.g., https://sitejs.org) that take a path to operate on:

```sh
site path/to/site/to/serve
```

If, say, eleventy was included in Site.js, we’d have to make sure that the only way the server was ever run was:

```sh
cd path/to/site/to/serve
site
```

(Unless I am missing something.)

**Comment by kleinfreund** [comment]
@aral Work in this MR is not blocking you from what you describe as it is not part of any Eleventy release so far. This MR has not been merged and may never get merged.

Is there an active issue you’re running into or is it rather that you anticipate this possibly becoming a problem in the future?

**Comment by MadeByMike** [comment]
Just wanted to pop in here because I had a legitimate where I wanted to use 11ty to serialise some data from collections that would be consumed by another service on my web server. It was actually a life-saver that 11ty allowed me to write a file outside the site root. All I needed was:
```
--- 
permalink: '../other-application/serialised-data.json'
---
```

I get the concerns, but also maybe 11ty is a low enough level tool that it should be able to do this? An SSG is essentially a build pipeline and it makes sense for these to be less opinionated about input and output. 

If this feature has to be removed I'd at least enable writing outside the root with a setting or flag. This will also ensure ongoing support for applications that make use of this feature.

**Comment by Ryuno-Ki** [comment]
I could live with a flag, that highlights, that this operation may be dangerous.

**Comment by zachleat** [comment]
Punting this out of 1.0—sorry!
